openreview/invitations/templates.py

```diff
@@ -232,4 +161,42 @@ def __init__(self, name, conference_id, duedate = 0,
         super(Comment, self).__init__(**self.inv_params)
 
     def add_process(self, process):
-        self.process = process.render()
\ No newline at end of file
+        self.process = process.render()
+
+def fill_template(template_string, paper_params):
+    pattern = '|'.join(['<{}>'.format(field) for field, value in paper_params.iteritems()])
+    matches = re.findall(pattern, template_string)
+    for match in matches:
+        discovered_field = re.sub('<|>', '', match)
+        template_string = template_string.replace(match, str(paper_params[discovered_field]))
+        print "    new value: ", template_string
+    return template_string
+
+def generate_invitation(invitation_template, paper):
+    params = copy.deepcopy(invitation_template)
+    reply = params.pop('reply')
+    paper_params = paper.to_json()
+
+    def fill_list_or_str(value, paper_params=paper_params):
+        if type(value) == list:
+            return [fill_template(v, paper_params) for v in value]
+        elif any([type(value) == t for t in [unicode, str]]):
+            return fill_template(value, paper_params)
+        elif any([type(value) == t for t in [int, float, type(None)]]):
+            return value
+        else:
+            raise ValueError('first argument must be list or string: ', value)
+
+    for field_map in [reply, params]:
+        for field in field_map:
+            print "field: ", field
+            if field == 'content' or field == 'process' or field == 'webfield':
+                pass
+            elif type(field_map[field]) == dict:
+                for subfield in field_map[field]:
+                    if subfield != 'description':
+                        field_map[field][subfield] = fill_list_or_str(field_map[field][subfield])
+            else:
+                field_map[field] = fill_list_or_str(field_map[field])
+
+    return openreview.Invitation(**dict(params, **{'reply': reply}))
```

Comment:
I would enforce the definition of each parameter when we create the invitation, and avoid the deep copy, for example:
```
openreview.Invitation(
writers = invitation_template[writers],
invitees = invitation_template[invitees],
....
)
```
We are sure that we are sending the correct parameters. Sending a dict of params you won't have control of which parameters are sending.

Reply:
actually, after thinking about this, I'm not sure if this is the right way to do it. invitations don't all have exactly the same set of fields, especially within the "reply" field. I think it should be the function caller's responsibility to set the parameters correctly - this function should only serve to iterate through the fields and replace the templates (e.g. <number>). This may suggest that this function should be named differently, like "fill_invitation_template()" instead of "generate_invitation()"

that being said, I'm interested to hear counter-arguments... what do you think?

Reply:
I disagree, we should define the contract and the function caller should respect it. This is the contract:

```
def __init__(self, id, writers=None, invitees=None, noninvitees=None, readers=None, nonreaders=None, reply=None, replyto=None, forum=None, invitation=None, web=None, process=None, signatures=None, duedate=None, cdate=None, rdate=None, ddate=None, multiReply=None, taskCompletionCount=None):

```

reply is a dictionary